crates/bevy_ecs/src/schedule/condition.rs

```diff
@@ -915,12 +921,104 @@ pub mod common_conditions {
     /// app.run(&mut world);
     /// assert_eq!(world.resource::<Counter>().0, 0);
     /// ```
-    pub fn not<Marker, T>(condition: T) -> impl Condition<()>
+    pub fn not<Marker, TOut, T>(condition: T) -> NotSystem<T::System, TOut>
     where
-        T: Condition<Marker>,
+        TOut: std::ops::Not,
+        T: IntoSystem<(), TOut, Marker>,
     {
-        condition.pipe(|In(val): In<bool>| !val)
+        let condition = IntoSystem::into_system(condition);
+        let name = format!("!{}", condition.name());
+        NotSystem::<T::System, TOut> {
+            condition,
+            name: Cow::Owned(name),
+        }
+    }
+}
+
+/// Invokes [`std::ops::Not`] with the output of another system.
+///
+/// See [`common_conditions::not`] for examples.
+#[derive(Clone)]
+pub struct NotSystem<T, TOut>
+where
+    TOut: std::ops::Not + 'static,
+    T: System<In = (), Out = TOut>,
```

Comment:
```suggestion
pub struct NotSystem<T: System>
where
    T::Out: std::ops::Not,
```

The `TOut` generic is unnecessary here, since we can just use the system's `Out` associated type. Also, we don't strictly need the input to be `()`.